test/unittest/tests/test_assertions.cpp

```diff
@@ -13,39 +13,335 @@
+{
+    WT_DECL_RET;
+
+    if (assert_should_pass)
+        WT_RET_PANIC_ASSERT(session, category, 1 == 1, -1, "WT_PANIC raised assert");
+    else
+        WT_RET_PANIC_ASSERT(session, category, 1 == 2, -1, "WT_PANIC raised assert");
+
+    ret = 14;
+
+    return ret;
+}
+
+/* Wrapper to call WT_ASSERT_OPTIONAL. */
+int
+call_wt_optional(WT_SESSION_IMPL *session, uint16_t category, bool assert_should_pass)
+{
+    if (assert_should_pass)
+        WT_ASSERT_OPTIONAL(session, category, 1 == 1, "WT_OPTIONAL raised assert");
+    else
+        WT_ASSERT_OPTIONAL(session, category, 1 == 2, "WT_OPTIONAL raised assert");
+
+    return 14;
+}
+
+/* Assert that WT_ASSERT and WT_ASSERT_ALWAYS behave consistently regardless of the HAVE_DIAGNOSTIC
+ * configuration. */
+int
+assert_always_aborts(WT_SESSION_IMPL *session)
+{
+    // WT_ASSERT does nothing.
+    WT_ASSERT(session, 1 == 2);
+    expect_no_assertion(session);
+
+    // WT_ASSERT_ALWAYS aborts.
+    WT_ASSERT_ALWAYS(session, 1 == 2, "Values are not equal!");
+    expect_assertion(session, "Assertion '1 == 2' failed: Values are not equal!");
+
+    return 0;
+}
+
+/* Assert that all diagnostic assert categories are off. */
+int
+all_diag_asserts_off(WT_SESSION_IMPL *session)
+{
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_DATA_VALIDATION)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_INVALID_OP)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_PANIC)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_CONCURRENT_ACCESS)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_OUT_OF_ORDER)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_SLOW_OPERATION)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_VISIBILITY)) == false);
+
+    return 0;
+}
+
+/* Assert that all diagnostic assert categories are on. */
+int
+all_diag_asserts_on(WT_SESSION_IMPL *session)
+{
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_DATA_VALIDATION)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_INVALID_OP)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_PANIC)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_CONCURRENT_ACCESS)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_OUT_OF_ORDER)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_SLOW_OPERATION)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_VISIBILITY)) == true);
+
+    return 0;
+}
+
+/* Assert that all asserts fire. */
+int
+all_asserts_abort(WT_SESSION_IMPL *session, uint16_t category, bool assert_should_pass)
+{
+    int ret = 0;
+
+    REQUIRE(assert_always_aborts(session) == 0);
+
+    REQUIRE(call_wt_optional(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_OPTIONAL raised assert");
```

Comment:
All of these asserts returning 14 feels a bit weird.. I think that means my suggestion of `#define ASSERT_NOT_FIRED 0` above won't work. Let's have a chat about what we can do here.